README.md
#Open TEC

Open Source Project (software and hardware) to get GPS data and calculate TEC ( Total Electron Content ) to be used in predicting the earthquake.

### Inspired by Prof. Kosuke Heki of Hokkaido University

<a href="http://www.youtube.com/watch?feature=player_embedded&v=yAVxsvE5ceM" target="_blank"><img src="https://dl.dropboxusercontent.com/u/79120458/kosuke-heki.jpg" 
alt="Earthquake Predictor using GPS" width="640" height="360" border="10" /></a>

Professor Kosuke Heki of Hokkaido University in Japan believes he has found a way to predict earthquakes.

Heki analyses GPS signals by measuring the TEC, or Total Electron Content, in the upper atmosphere. Whilst measuring how the TEC was disrupted by sound waves after the Tohoku earthquake of 2011, he discovered - quite by accident - that the TEC was also disrupted in the 40 or so minutes before it.

Going back through GPS records he has found similar correlations for other major earthquakes, a discovery that is being heralded as a major breakthrough in our understanding of earthquake phenomena.

##### His Abtract:

The Japanese dense network of Global Positioning System (GPS) detected preseismic enhancement of ionospheric total electron content (TEC) over the region within 200-300 km from the focal region. It started ~60-40 minutes before the 2011 March 11 NE Japan earthquake (Mw 9.0) , and reached ~8 % of the background TEC. It lasted until atmospheric waves arrived at the ionosphere. Similar preseismic TEC anomalies, with amplitudes dependent on magnitudes, were seen in the 2010 Chile (Mw 8.8), the 2004 Sumatra-Andaman (Mw 9.2), the 2007 Bengkulu (Mw 8.6), and 
the 1994 Hokkaido-Toho-Oki (Mw 8.3) earthquakes, but not before earthquakes with Mw 8 or less.

### Paper

<a href="http://www.ep.sci.hokudai.ac.jp/~heki/pdf/Heki_GRL2011.pdf" target="_blank">Heki, K., Ionospheric electron enhancement preceding the 2011 Tohoku-Oki earthquake, Geophys. Res. Lett. 38, L17312, doi:10.1029/2011GL047908, 2011.</a>

### Poster

<a href="http://www.ep.sci.hokudai.ac.jp/~heki/pdf/2011AGU_Poster.pdf" target="_blank">Precursory TEC enhancement immediately before the 2011 Tohoku-Oki earthquake.</a> 

### Open TEC implementation:

```
1. Get Raw GPS Data from Arduino + Itead Arduino GPS Shield in SiRF binary 
2. Convert SiRF binary to RINEX using 'sirfdump' 
3. Calculate TEC using 'gpstk'
```

### Installation

##### sirfdump

```
make
sudo make install
```

##### gpstk

System Requirements: (<a href='http://www.gpstk.org/bin/view/Documentation/SystemRequirements' target='_blank'>link</a>)

```
tar xvzf gpstk.tar.gz
cd gpstk/dev
./autogen.sh
./configure
make
sudo make install
```

##### GPS logger

```
Use gps_logger.pde with Arduino IDE and upload it to the board to store SiRF Binary output into MiniSD card. 
```


gpstk/NEWS
$Id: NEWS 3140 2012-06-18 15:03:02Z susancummins $

GPS Toolkit News
------------------------------------------------------------------------------
   June 18, 2012 . . . .  Release Candidate for Version 2.0 is frozen in the 
                          repository

   February 14, 2011 . .  Release Candidate for Version 1.7 is frozen in the
                          repository
   
   June 22, 2009 . . . .  Release Candidate for Version 1.6 is frozen in the 
			  repository

   July 16, 2007 . . . .  Release Candidate for Version 1.3 is frozen in the 
			  repository

   January 10, 2005 . . . Version 1.1 is released.

   September 20, 2004 . . Version 1.0.2 is released to coincide with
                          presentation at the ION GNSS 2002 in
                          Portland, Oregon.

   August 9, 2004 . . . . Version 1.0.1 released quickly to address
                          minor issues.

   August 8, 2004 . . . . The first version of the GPSTk is released to 
                          coincide with an article in LinuxJournal.

   February 25, 2004  . . The GPSTk project is registered at SourceForge.net.


gpstk/src/Xvt.cpp
#pragma ident "$Id: Xvt.cpp 3140 2012-06-18 15:03:02Z susancummins $"

/**
 * @file Xvt.cpp
 * Position and velocity as Triples, clock bias and drift as doubles.
 */

//============================================================================
//
//  This file is part of GPSTk, the GPS Toolkit.
//
//  The GPSTk is free software; you can redistribute it and/or modify
//  it under the terms of the GNU Lesser General Public License as published
//  by the Free Software Foundation; either version 2.1 of the License, or
//  any later version.
//
//  The GPSTk is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU Lesser General Public License for more details.
//
//  You should have received a copy of the GNU Lesser General Public
//  License along with GPSTk; if not, write to the Free Software Foundation,
//  Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110, USA
//  
//  Copyright 2004, The University of Texas at Austin
//
//============================================================================

//============================================================================
//
//This software developed by Applied Research Laboratories at the University of
//Texas at Austin, under contract to an agency or agencies within the U.S. 
//Department of Defense. The U.S. Government retains all rights to use,
//duplicate, distribute, disclose, or release this software. 
//
//Pursuant to DoD Directive 523024 
//
// DISTRIBUTION STATEMENT A: This software has been approved for public 
//                           release, distribution is unlimited.
//
//=============================================================================

#include <iostream>
#include "Xvt.hpp"

// Output operator for Xvt
// @param os output stream to which \c xvt is sent
// @param xvt Xvt that is sent to \c os
std::ostream& operator<<(std::ostream& os, const gpstk::Xvt& xvt) throw()
{
   os << "x:" << xvt.x
     << ", v:" << xvt.v
     << ", clk bias:" << xvt.clkbias
     << ", clk drift:" << xvt.clkdrift
     << ", relcorr:" << xvt.relcorr;
   return os;
}

// compute the relativity correction
double gpstk::Xvt::computeRelativityCorrection(void)
{
   relcorr = -2.0*( (x[0]/C_MPS)*(v[0]/C_MPS)
                   +(x[1]/C_MPS)*(v[1]/C_MPS)
                   +(x[2]/C_MPS)*(v[2]/C_MPS) );
   return relcorr;
}

// Function to find the range and position from a ground
// location, rxPos, to the spacecraft position (*this).x
// Use the pseudorange corrected for SV clock effects to get a
// rough time of flight (dt).  Account directly for Earth
// rotation, then compute a rough receiver bias by differencing
// the initial time of flight with the new estimate.  Then
// correct the rotation by a small amount.
double gpstk::Xvt::preciseRho(const Triple& rxPos,
                              const EllipsoidModel& ellips,
                              double correction) const 
   throw()
{
   // Compute initial time of flight estimate using the
   // geometric range at transmit time.  This fails to account
   // for the rotation of the earth, but should be good to
   // within about 40 m
   double sr1 = rxPos.slantRange(x);
   double dt = sr1 / ellips.c();

   // compute rotation angle in the time of signal transit
   double rotation_angle = -ellips.angVelocity() * dt;

   // rotate original GS coordinates to new values to correct for
   // rotation of ECEF frame
   // Ref: Satellite Geodesy, Gunter Seeber, 1993, pg 291  and the 
   // ICD-GPS-200 sheet 102 May 1993 version
   //   xnew[0]=xg[0]*cos(rotation_angle)-xg[1]*sin(rotation_angle);
   //   xnew[1]=xg[1]*cos(rotation_angle)+xg[0]*sin(rotation_angle);
   //   xnew[2]=xg[2];
   // since cosine and sine are small, approximate by the first
   // order terms in an expansion.
   gpstk::Triple xnew;
   for (int i = 0; i < 2; i++)
   {
      xnew[0] = x[0] - x[1] * rotation_angle;
      xnew[1] = x[1] + x[0] * rotation_angle;
      xnew[2] = x[2];

      // Compute geometric slant range from ground station to
      // the rotated new coord's
      sr1 = rxPos.slantRange(xnew);
   
      // Recompute the time of flight (dt) based on PR, with the
      // time of flight based on geometric range.  Note that
      // this is a really unneeded, in that the change in PR is
      // < 40 m, hence the change in tof is < 20 ns
      dt = sr1 / ellips.c();
   
      // Compute new rotation in this time 
      rotation_angle = -ellips.angVelocity() * dt;  
   }

   // Account for SV clock drift, relativity and user input correction
   double rho = sr1 - (clkbias + relcorr) * ellips.c() - correction;

   return rho;

} // end of preciseRho()

